File: Maximize Area of Square Hole in Grid/Maximize Area of Square Hole in Grid.cpp
// Solution for Maximize Area of Square Hole in Grid in CPP

class Solution {
public:
    /*
        Finds the maximum number of consecutive removable bars.
        Example: [2,3,4] → 3 consecutive bars
    */
    int maxConsecutiveBars(vector<int>& bars) {
        if (bars.empty()) return 0;

        int maxCount = 1;
        int currentCount = 1;

        for (int i = 1; i < bars.size(); i++) {
            if (bars[i] == bars[i - 1] + 1) {
                currentCount++;
            } else {
                currentCount = 1;
            }
            maxCount = max(maxCount, currentCount);
        }
        return maxCount;
    }

    int maximizeSquareHoleArea(int n, int m, vector<int>& hBars, vector<int>& vBars) {

        // Sort bars to detect consecutive removals
        sort(hBars.begin(), hBars.end());
        sort(vBars.begin(), vBars.end());

        int maxHorizontal = maxConsecutiveBars(hBars);
        int maxVertical = maxConsecutiveBars(vBars);

        /*
            Removing k consecutive bars creates (k + 1) free cells
            Square side = min(horizontalCells, verticalCells)
        */
        int squareSide = min(maxHorizontal, maxVertical) + 1;

        return squareSide * squareSide;
    }
};

File: Maximize Area of Square Hole in Grid/Maximize Area of Square Hole in Grid.py
# Solution for Maximize Area of Square Hole in Grid in PY

class Solution:
    """
    Finds the maximum number of consecutive removable bars.
    Example: [2,3,4] → 3 consecutive bars
    """
    def maxConsecutiveBars(self, bars):
        if not bars:
            return 0

        max_count = 1
        current_count = 1

        for i in range(1, len(bars)):
            if bars[i] == bars[i - 1] + 1:
                current_count += 1
            else:
                current_count = 1
            max_count = max(max_count, current_count)

        return max_count

    def maximizeSquareHoleArea(self, n, m, hBars, vBars):
        # Sort bars to detect consecutive removals
        hBars.sort()
        vBars.sort()

        max_horizontal = self.maxConsecutiveBars(hBars)
        max_vertical = self.maxConsecutiveBars(vBars)

        # Removing k consecutive bars creates (k + 1) free cells
        square_side = min(max_horizontal, max_vertical) + 1

        return square_side * square_side
